include/dr/mhp/algorithms/inclusive_scan.hpp

```diff
@@ -0,0 +1,275 @@
+
+  auto d_last = d_first;
+  rng::advance(d_last, rng::size(r));
+
+  // TODO: current alg assumes local segments are contiguous and located
+  //  in rank order. This is OK as all data structures we have currently have
+  //  just one segment per rank and in rank order.
+  auto in_segments = local_segments(r);
+  auto out_segments = local_segments(rng::subrange(d_first, d_last));
+
+  // rng::size(in_segments) doesn't work and can't count non-empty segments
+  std::size_t local_segments_count = 0;
+  for (auto &&s [[maybe_unused]] : in_segments)
+    if (!rng::empty(s))
+      ++local_segments_count;
+
+#ifdef SYCL_LANGUAGE_VERSION
+  std::vector<sycl::event> events;
+#endif
+
+  using OVal = rng::iter_value_t<O>;
+  // how to get allocator used with input and output range? I assume it a
+  // default one probably we should forbid using non-default ones with
+  // distributed_ranges containers (like distributed_vector)
+  OVal *local_partial_sums =
+      local_segments_count
+          ? default_allocator<OVal>().allocate(local_segments_count)
+          : nullptr;
+  OVal *local_partial_sums_iter = local_partial_sums;
+
+  for (auto &&segs : rng::views::zip(in_segments, out_segments)) {
+
+    auto &&[in_segment, out_segment] = segs;
+
+    auto dist = rng::distance(in_segment);
+    if (dist == 0) {
+      continue;
+    }
+
+    auto first = ranges::local(rng::begin(in_segment));
+    auto last = ranges::local(rng::end(in_segment));
+    auto s_first = ranges::local(rng::begin(out_segment));
+
+#ifdef SYCL_LANGUAGE_VERSION
+    sycl::event one_segment_scan_event;
+    if (mhp::use_sycl()) {
+      one_segment_scan_event = oneapi::dpl::experimental::inclusive_scan_async(
+          dpl_policy(), dr::__detail::direct_iterator(first),
+          dr::__detail::direct_iterator(last),
+          dr::__detail::direct_iterator(s_first), binary_op);
+    } else
+#endif
+      std::inclusive_scan(std::execution::par_unseq, first, last, s_first,
+                          binary_op);
+
+    rng::advance(s_first, dist - 1);
+
+#ifdef SYCL_LANGUAGE_VERSION
+    if (mhp::use_sycl()) {
+      events.push_back(mhp::sycl_queue().submit([&](auto &&h) {
+        h.depends_on(one_segment_scan_event);
+        h.single_task([=]() {
+          OVal tmp = *s_first;
+          *local_partial_sums_iter = tmp;
+        });
+      }));
+    } else
+#endif
+      *local_partial_sums_iter = *s_first;
+    ++local_partial_sums_iter;
+  }
+
+#ifdef SYCL_LANGUAGE_VERSION
+  sycl::event::wait(events);
+  events.clear(); // events are reused later in this function
+#endif
+
+  const std::size_t local_partial_sums_count =
+      rng::distance(local_partial_sums, local_partial_sums_iter);
+
+  OVal *local_partial_sums_scanned =
+      local_partial_sums_count
+          ? default_allocator<OVal>().allocate(local_partial_sums_count)
+          : nullptr;
+
+  if (local_partial_sums_count) {
+#ifdef SYCL_LANGUAGE_VERSION
+    if (mhp::use_sycl())
```

Comment:
oneapi coding standard suggests always using `{}`, even when `if/else` has 1 statement. Using `{}` and moving `ifdef` into them would make the code easier to read.

Reply:
Braces - done. `ifdef`s left outside but moved to separate functions.